### ISSUE TITLE
Automated profiling of user-submitted bug reports

### ISSUE THREAD

**Issue by ehildenb** [runtimeverification/haskell-backend#3221]
A given bug report has all the needed information to reproduce the execution of the haskell backend. There are standard ways we have of profiling these executions.

See here: https://github.com/runtimeverification/haskell-backend/issues/3216#issuecomment-1229193911

In particular, we have a way to setup cabal to do this type of profiling: https://github.com/runtimeverification/haskell-backend/wiki/Profiling#haskell-profiling. And we also have a way to do memory profiling: https://github.com/runtimeverification/haskell-backend/blob/master/docs/profiling.md

We should setup a CI job that allows:

- Only Run performance test ONLY when submitted by RV Employees.
- Create issue template for Performance related Issues -- Used to trigger performance CI w/ given .tar file. 
- User to submit a bug report as a GitHub issue, with the tarball attached to the main issue body.
- CI machine grabs that tarball, and builds haskell backend with profiling turned on, then profiles the bug report.
- Uploads all the generated artifacts directly to the github issue (speedscope logs, etc...).

As we improve our profiling setup over time, we can make sure this harness is also taking advantage of those changes. This will hopefully make some small portion of the issues much quicker to fix. Maybe we can immediately spot the problem in the initial profiles, without having to do much extra work, and this removes the human step of having to download and run those profiles.

**Comment by ehildenb** [comment]
@F-WRunTime can help with setting up the CI side of things.

**Comment by ana-pantilie** [comment]
@jberthold @goodlyrottenapple @ana-pantilie please add a comment describing how you generate profiles and what info you'd like to get automatically when someone opens an issue like this.

**Comment by ana-pantilie** [comment]
### An overview of my workflow when first profiling a bug report

Requirements: cabal, [hs-speedscope](https://github.com/mpickering/hs-speedscope)

With a fresh copy of the project:
```
git clone git@github.com:runtimeverification/haskell-backend.git
cabal v2-configure --enable-profiling -f-threaded
cabal v2-build all
// to successfully automate this the newly built kore-exec should be added to path somehow
cd <path-to-bug-report>
timeout -s INT 4m env GHCRTS='-p -l-au' ./kore-exec.sh
hs-speedscope kore-exec.eventlog
```
That should output `kore-exec.eventlog.json` which can be loaded into https://www.speedscope.app/ and viewed.

**Comment by jberthold** [comment]
For convenience, I am usually using `stack install --profile` and have the binary output directory on the path.

With `GHCRTS='-p'` alone, one gets a textual time/memory profile which is what I usually use in the first place.
I haven't worked much with speedscope yet but I guess it is useful for _partial_ profile data from a timed-out program run.
According to the speedscope manual, it can import GHC profiles created with `GHCRTS='-pj' directly but this RTS option is only for GHC>=9.0 (and I haven't tried it)

**Comment by goodlyrottenapple** [comment]
I have set up some targets in nix for profiling, especially for GHC9 which has some new profiling features. I usually do
```
nix shell github:runtimeverification/haskell-backend#kore-exec-prof-ghc9
```
or 
```
nix shell .#kore-exec-prof-ghc9
```
if i am already in the haskell-backend repo and then something similar to @ana-pantilie's
```
timeout -s INT 4m env GHCRTS='-p -l-au' ./kore-exec.sh
```

**Comment by ana-pantilie** [comment]
We need to write a script which builds the latest master with profiling and generates:
- textual/json profile
- eventlog profile
- standard heap profile (cost centres)

Can we do the above in the same run? Will the eventlog profile contain different information depending on the other profiling options?

Can we view the html files directly from eventlog2html? Should we just post the picture of the graph in the issue?

Also set up heap profiling with GHC 9. Needs a different build, but not a profiling build (eventlog needs to be turned on but it's already on by default on our project). This needs to be checked, if it's true then merge https://github.com/runtimeverification/haskell-backend/pull/3231. We want to do this type of profiling: https://well-typed.com/blog/2021/01/first-look-at-hi-profiling-mode/

Test this on #3216 

We will want to run this on our own machines because we need the profiles to contain as much information as possible.

**Comment by ehildenb** [comment]
We want to be able to run it anywhere, which means that it should be incorporated into the build system committed to the repo.

But we also want a "default" template, with a series of simple-to-run commands, which can be run on CI, so that we can automate the first round of collecting that performance information. That way, without having to do anything, you can already have a discussion as a team about those initial profiles that are produced automatically, before assigning more specific investigations.

**Comment by ehildenb** [comment]
Who will take this? I think it's high priority @ana-pantilie. Collecting profiles should be as low-friction as possible.

**Comment by ehildenb** [comment]
We should run these jobs on an isolated instance. We can tag each instance with `performance` tag, and use the old rv-master node (which should be unused now) with a single github runner taking up 95% of the capabilities of the machine.

**Comment by F-WRunTime** [comment]
Got First tests Executing here: https://github.com/runtimeverification/devops-actions/actions/runs/3063300591/jobs/4945220719

**Comment by F-WRunTime** [comment]
Further information How to use: 
When using in a pull request a performance test can be triggered via being labeled properly FIRST, then adding a comment with to the PR with the .tar file


![Image](https://user-images.githubusercontent.com/105403280/190525125-afa3c596-31df-4b44-85ff-91f7b8ba6d80.png)

Alternatively can be triggered when adding a comment to an existing or in the issue body. 
Editing an existing Issue or PR triggers the same by adding a comment and the tar file. 

![Image](https://user-images.githubusercontent.com/105403280/190525214-9d11d4d1-101c-4a5f-8786-a8732094fe09.png)

Label MUST be 'performance' on the issue or pull request for a test to run. Otherwise the trigger will skip and not run. 

Tests currently are run against the primary branch only.

**Comment by F-WRunTime** [comment]
And it was working when I posted it ran a test again and it complains about an error that didn't exist before. 
Further investigation needed...

**Comment by F-WRunTime** [comment]
Made some progress this morning. Got a couple runs setup to go now. Made a fix with suggestion from @jberthold 
Fixed an issue in the first run with thefile name. Think the Environment file is only sourced once per step (makes sense). So should see a couple successful run throughs here. Hopefully this means we can look to move this over to the haskell repository in a PR on Monday and get some auto profiling running! 
@ehildenb @ana-pantilie @goodlyrottenapple let me know if you have any questions 
changes are currently on master of the [devops-actions repository ](https://github.com/runtimeverification/devops-actions/actions/runs/3069572010)
I will remove concurrency for the haskell repo.

**Comment by F-WRunTime** [comment]
@goodlyrottenapple[ Any insight as to why this is failing](https://github.com/runtimeverification/devops-actions/actions/runs/3069572010/jobs/4958385049)

It says it passes but on step: "Run Tests" it is running the 'nix command' we discussed
And there are errors as shown below. 

```bash
gzip: stdin: unexpected end of file
/nix/store/mnln1nxdhxx5qj8rcyvllhhs6d4h668b-gnutar-1.34/bin/tar: Child returned status 1
/nix/store/mnln1nxdhxx5qj8rcyvllhhs6d4h668b-gnutar-1.34/bin/tar: Error is not recoverable: exiting now
env: ‘./kore-exec.sh’: No such file or directory
cp: cannot stat 'kore-exec.prof': No such file or directory
cp: cannot stat 'kore-exec.eventlog': No such file or directory
hs-speedscope: kore-exec.eventlog: openBinaryFile: does not exist (No such file or directory)
cp: cannot stat 'kore-exec.eventlog.json': No such file or directory
env: ‘./kore-exec.sh’: No such file or directory
cp: cannot stat 'kore-exec.hp': No such file or directory
cp: cannot stat 'kore-exec.eventlog': No such file or directory
hp2pretty: kore-exec.hp: openFile: does not exist (No such file or directory)
eventlog2html: kore-exec.eventlog: openBinaryFile: does not exist (No such file or directory)
cp: cannot stat 'kore-exec.svg': No such file or directory
cp: cannot stat 'kore-exec.eventlog.html': No such file or directory
env: ‘./kore-exec.sh’: No such file or directory
cp: cannot stat 'kore-exec.eventlog': No such file or directory
eventlog2html: kore-exec.eventlog: openBinaryFile: does not exist (No such file or directory)
cp: cannot stat 'kore-exec.eventlog.html': No such file or directory
/home/github-runner/actions-runner/_work/_temp/0daef5be-08b4-4580-b7[91](https://github.com/runtimeverification/devops-actions/actions/runs/3069572010/jobs/4958385049#step:4:92)-ad4b673a6d8e.sh: line 13: Fri Sep 16 17:37:17 UTC 2022: syntax error in expression (error token is "Sep 16 17:37:17 UTC 2022")
```
The current snapshot of the _work folder looks as follows: 
```bash
github-runner@github-runner-1:~/actions-runner/_work/devops-actions/devops-actions$ ls -l
total 32
-rw-r--r-- 1 github-runner github-runner    0 Sep 16 17:22 foundry.tar.gz
drwxr-xr-x 3 github-runner github-runner 4096 Sep 15 21:11 profile-20220915-211153-foundry
-rw-r--r-- 1 github-runner github-runner  306 Sep 15 21:11 profile-20220915-211153-foundry.tar.gz
drwxr-xr-x 3 github-runner github-runner 4096 Sep 15 22:58 profile-20220915-225804-foundry
-rw-r--r-- 1 github-runner github-runner  307 Sep 15 22:58 profile-20220915-225804-foundry.tar.gz
drwxr-xr-x 3 github-runner github-runner 4096 Sep 15 23:02 profile-20220915-230241-foundry
-rw-r--r-- 1 github-runner github-runner  308 Sep 15 23:02 profile-20220915-230241-foundry.tar.gz
drwxr-xr-x 3 github-runner github-runner 4096 Sep 16 17:37 profile-20220916-173717-foundry
-rw-r--r-- 1 github-runner github-runner  307 Sep 16 17:37 profile-20220916-173717-foundry.tar.gz

```
Is there a known issue with having multiple runs in the same folder? 
Only issue I see is the artifact uploads after finished are uploading all the runs. I will add a 'clean folders' before starting a new run to make sure there is no left overs from previous runs.

**Comment by jberthold** [comment]
Looking at https://github.com/runtimeverification/haskell-backend/blob/master/nix/run-profiling.nix#L15 (which seems to again be the first failure in the script), and your directory listing:
`-rw-r--r-- 1 github-runner github-runner    0 Sep 16 17:22 foundry.tar.gz`
This is an existing empty file. On my machine, `tar -xvzf` will respond with the error we see, `gzip: stdin: unexpected end of file` when given an existing empty file.

Also, we should definitely add a `-e` to the shell script that runs the profile. No point trying anything further if unwrapping failed. Just need to ensure the steps that _may_ fail (like a `timeout` call)  are wrapped so we still get results.

And yes, cleaning up the working directory would be a good idea. I am not sure but the script might also have an issue [confusing `./profile` and `$out_folder`](https://github.com/runtimeverification/haskell-backend/blob/master/nix/run-profiling.nix#L12-L15).

**Comment by F-WRunTime** [comment]
@jberthold 

Added some fixes but when I run the test I'm still getting some errors, here is me running it manually with the foundry.tar.gz test file 
```bash
github-runner@github-runner-1:~/actions-runner/_work/devops-actions/devops-actions$ nix run github:runtimeverification/haskell-backend#profile ./foundry.tar.gz
kore-exec.sh
vdefinition.kore
spec.kore
kore-exec.log
result.kore
tests/foundry/foundry.debug-log: openFile: does not exist (No such file or directory)
tests/foundry/foundry.debug-log: openFile: does not exist (No such file or directory)
/tmp/kore-exec-e82bd9f554ffa191/kore-exec.log: withFile: does not exist (No such file or directory)
```

**Comment by F-WRunTime** [comment]
@goodlyrottenapple @jberthold 
Please see latest results. I am getting an error now regarding 1000 traces. 
https://github.com/runtimeverification/devops-actions/actions/runs/3084571476/jobs/4986872128

Is there a configuration I'm missing?

**Comment by jberthold** [comment]
The `1000 traces` is actually a warning, emitted by the tool that post-processes a trace log into speedscope format. The tool has options to filter the trace log for specific K rules, and speedscope might have trouble loading data which has too many traces.
What errored out in the run linked above was the computation on dates (`Mon` for `Monday` , the time when this ran...)

**Comment by F-WRunTime** [comment]
> 

Okay good to know. Sounds like it's fine though just to be clear before leaving it alone?

**Comment by ana-pantilie** [comment]
Fixed by #3265
